core/drivers/crypto/caam/caam_key.c

```diff
@@ -0,0 +1,739 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright 2023 NXP
+ */
+#include <assert.h>
+#include <caam_desc_helper.h>
+#include <caam_key.h>
+#include <caam_status.h>
+#include <caam_trace.h>
+#include <caam_utils_mem.h>
+#include <crypto/crypto.h>
+#include <kernel/panic.h>
+#include <mm/core_memprot.h>
+#include <stdint.h>
+#include <string.h>
+#include <tee/cache.h>
+#include <tee_api_defines.h>
+#include <trace.h>
+#include <utee_types.h>
+
+/*
+ * CAAM Key magic number.
+ * This 32 bits value added to the first 32 bits tells the CAAM driver if
+ * the given key buffer is a serialized CAAM key structure.
+ */
+#define MAGIC_NUMBER 0xCAAFBFFB
+
+#define BLACK_KEY_NONCE_SIZE 6	/* CCM Black key nonce size */
+#define BLACK_KEY_ICV_SIZE   6	/* CCM Black key ICV size */
+#define BLOB_BKEK_SIZE	     32 /* Blob key encryption key size */
+#define BLOB_MAC_SIZE	     16 /* Blob MAC size */
+#define BLOB_PAD_SIZE	     (BLOB_BKEK_SIZE + BLOB_MAC_SIZE)
+
+/*
+ * CAAM Blob key modifier
+ * Key modifier used to derive Blob-key encryption key (BKEK) from the CAAM
+ * master key.
+ *
+ * A CAAM black key is encrypted using a volatile Job Descriptor key encryption
+ * key or JDKEK. Black keys are not intended for storage of keys across SoC
+ * power cycles. The JDKEK is re-generated upon every power cycle (reset,
+ * suspend/resume ...) or CAAM RNG re-seed.
+ *
+ * To retain key across power cycles, the black key must be encapsulated as a
+ * blob. The blob key encryption key is derived from the CAAM master key which
+ * makes it non-volatile and can be re-created when the chip powers up again.
+ */
+#define KEY_BLOB_MODIFIER_SIZE 16
+static const uint8_t key_blob_modifier[KEY_BLOB_MODIFIER_SIZE] =
+	"NXP_OPTEE_BLOB";
+
+/*
+ * Serialized CAAM key structure format.
+ *
+ * If the incoming key buffer is the following:
+ *	| Magic number | key type | key size | key blob buffer |
+ * The CAAM Key structure will be populated as following:
+ * struct caamkey {
+ *	.key_type = key type,
+ *	.key_size = key size,
+ *	.is_blob = true,
+ *	.buf = key blob buffer
+ * }
+ *
+ * If the incoming key buffer is the following:
+ *	| Key buffer |
+ * The CAAM Key structure will be populated as following:
+ * struct caamkey {
+ *	.key_type = CAAM_KEY_PLAIN_TEXT,
+ *	.key_size = sizeof(Key buffer),
+ *	.is_blob = false,
+ *	.buf = key buffer
+ * }
+ */
+struct caam_key_serialized {
+	struct {
+		uint32_t magic_number; /* Magic number */
+		uint32_t key_type;     /* Black key type */
+		uint32_t sec_size;     /* The original plain text key size */
+	} header;
+	uint8_t key[];
+} __packed;
```

Comment:
The proper way to do that is to make a struct layout that is free from padding by nature and then add a `static_assert()` to make sure that there really isn't any unexpected padding. Dropping the inner struct should be sufficient.

That leads to more efficient machine code, better warnings, and also saves some time for maintainers since there's less chance that others copy the slightly flawed `__packed` pattern. :-)

Reply:
If I drop `__packed` I will get a cast warning
```

core/drivers/crypto/caam/caam_key.c: In function ‘data_to_serialized_key’:
core/drivers/crypto/caam/caam_key.c:99:15: warning: cast increases required alignment of target type [-Wcast-align]
   99 |         key = (struct caam_key_serialized *)data;
      |               ^
core/drivers/crypto/caam/caam_key.c: At top level:
cc1: note: unrecognized command-line option ‘-Wno-c2x-extensions’ may have been intended to silence earlier diagnostics
```
A static assert can make sure it won't happen, but I will still get the warning, right?

Reply:
You can cast it to `void *` instead since you have checked that the address is well aligned.

Reply:
I would personally go with either `__packed` or a `memcpy()`. I find it tricky to make a `void *` cast dependent of `assert()` in another line of code somewhere else.
I guess if we go that way, it must be properly documented in the code.

Reply:
Feel free to add a comment.

Reply:
On?

Reply:
> I guess if we go that way, it must be properly documented in the code.

Reply:
https://github.com/OP-TEE/optee_os/pull/5990/files#diff-2af92fa52b0087bf848ecd8b9b9de69c7330aa1b63cb31c3388387e1dcdb5bd7R109

Reply:
I'm sorry, I don't understand, what are you proposing?

Reply:
Oops sorry, wrong link. I mean, I went with the `(void *)` cast and add comments : https://github.com/OP-TEE/optee_os/blob/21b78605d29a8a1d62f4cc483607ae429ccf7e94/core/drivers/crypto/caam/caam_key.c#L109